### ISSUE TITLE
A general question on multi-agent reinforcement learning

### ISSUE THREAD

**Issue by kwea123** [Unity-Technologies/ml-agents#155]
Hello,

First of all thanks for your great work on this project. I would like to ask a question about multi-agent reinforcement learning as one of your environment, Tennis, actually employs this scenario.

From what I understand, in multi-agent scenarios, we are trying to find (or to approach) a Nash equilibrium (NE) of the game. In case we have only two adversarial agents against each other, we can model the scenario as a two player zero sum game, by setting rewards such that when one agent gets a positive reward, the other gets a negative one with the same absolute value. There are many interesting properties of these kind of games that are proven, for example that a mixed strategy NE always exists, and all NE have the same payoff values.

To find a such NE, we can use self plays (like in Alphago), and finally we get a (approximated) mixed strategy NE profile and the (also approximated) payoff matrix.

My question is, is my understanding correct, that in multi-agent scenarios (and in self plays) we're trying to find out each player's strategy profile in a mixed strategy NE (which is what the policy network does), and the value network corresponds to the payoff matrix?

Also, If so, how do we deal with continuous actions? In discrete action space we can approximate all sorts of distributions, but in continuous space, we usually model the action with a gaussian distribution, which is very restrictive here. For example if the mixed NE strategy requires 1/2 probability of playing "1" and 1/2 of playing "0", then as far as I know, no traditional distributions can even approach this.

Any feedback is appreciated. Thank you.

**Comment by kwea123** [comment]
Oh... It doesn't even work on extremely simple things...
I'm using a very simple network PPO to train matching pennies game(https://en.wikipedia.org/wiki/Matching_pennies) agents. This game has a mixed strategy NE, which is 1/2 proba playing heads and 1/2 proba playing tails for both players.

The agent works well if the other player is using a deterministic policy (e.g. if the other only plays heads or only plays tails), but when I tried using self plays, the result doesn't look good...
![1](https://user-images.githubusercontent.com/11364490/33229676-1a459f8c-d20e-11e7-8fde-2b1c0aff8048.png)

Basically, you can see that when P1(0), the probability of player 1 choosing face 0, decreases, P2(0) decreases too, which is not reasonable, since in my setup player 2 wins when the players play differently. I expected the result P1(0) to be like a sin(x)/x function (for x>0),
![2](https://user-images.githubusercontent.com/11364490/33229836-25804b00-d212-11e7-97af-7f1e6e67c8a1.gif)
which oscillates around 0.5, which means the players constantly adapt to the other's strategy.

I wonder if your how your Tennis agents work actually, and how's their performance. Thank you.

**Comment by kwea123** [comment]
I did some experiments on simple games (not related to unity), results are as follows:

1. Matching pennies (https://en.wikipedia.org/wiki/Matching_pennies)
The mixed strategy NE is both player playing each face with probability 0.5
![1](https://user-images.githubusercontent.com/11364490/33230794-e1cac72e-d224-11e7-96d3-a9120d4ebf51.png)

2. Paper scissor stone
The mixed strategy NE is both player playing each move with probability 1/3 = 0.333
![2](https://user-images.githubusercontent.com/11364490/33230818-283c8968-d225-11e7-851c-7d6759ab4265.png)

**Comment by vincentpierre** [comment]
Can you look at the rewards ? Are the rewards as high as you expect ? Can you do better than the trained agent if you implement the heuristic brain?   
Since this game is mixed strategies, I think it would not necessarily converge to the optimal since the agents have multiple attempts. It is not a one time game in this setting, but I have not looked into this game in details. Your stuff looks amazing!

**Comment by kwea123** [comment]
Hi, I found PPO not working at all here, the images I got was totally coincidence. I did some research, and I tried out the most simple solution, which is trying to maximize the expected rewards.

I use a single neural network to be used by the agents, which takes the state as input and output the (discrete space) actions' probabilities. Then I feed the all action combinations and the corresponding rewards (i.e. the payoff matrix). Each agent (for now there are only two) modifies his strategy according to the other's and updates the neural network.

I proved mathematically that this is going to converge to the NE strategy in a e^(-x)cos(x) way, and after trying on multiple different games, including matching pennies, I found this working quite well. Following is an image showing the "tweaked" matching pennies training :
![1](https://user-images.githubusercontent.com/11364490/33330386-b51dfc56-d499-11e7-915b-b12b5abacb50.png)
It converges well to the NE, where each player should play face 0 with proba 2/5=0.4 as I stated previously.

In games where each player has more than two choices, this also works theoretically, although I've only tried on normal paper-scissors-stone game so far. Following is the training process :
![12](https://user-images.githubusercontent.com/11364490/33330604-3cd7b524-d49a-11e7-8ead-b1a4467c783a.png)

This also converges in e^(-x)cos(x) way and the optimal value is 1/3.

However, I noticed that it's somehow difficult to always get the convergence, although mathematically proven; I need to tune the learning rate and the number of updates correctly, otherwise it can only oscillate around the optimal value and never converge.

Finally, I think the reason why PPO doesn't work here is that the game is too small here, so the critic network quickly converges to what is not optimal; and since we encourage exploration by adding an entropy term, we encourage the proba distribution to be equally distributed (so 1/2 1/2 in two-choices games). The quick convergence of the critic network and the encouragement exploration make the advantage and ratio = pi/old_pi small respectively, and the not-moving strategy further accelerates the convergence of the critic net. As a result it can only learn the equal probability strategy (every choice with the same probability), and it thinks that the reward is the best. That's why after tweaking the rewards, it still doesn't move away from 1/2 1/2 proba distribution.

I wonder how's your opinion, and if you have any suggestion on papers that I can read; at the same time, I want to try my method on tweaked paper-scissors-stone to see if it still converges, and maybe try to add a third agent, and so on... I will keep you updated if I have any progress.

Here's my ipynb on this little project:
https://github.com/kwea123/RL/blob/master/ai/unity_test/matching_pennies/two_player_two_option_games.ipynb

Thanks for reading.
